File: Data Structures & Algorithms/climbing-stairs/submission-4.java
class Solution {
    public int count;
    public int climbStairs(int n) {
        //use rescursive tree
        // find the memo approach
        count = 0;
        
        helper(0,n);

        return count;
    }

    public void helper(int curr, int n){
        //break
        if(curr == n){
            count++;
            
            return;
        }

        if (curr > n ) return;

        //pick 1
        
        helper(curr + 1, n);
        //dont pick 1 (2)
        helper(curr + 2, n);

    }
}

File: Data Structures & Algorithms/subsets/submission-4.cs
public class Solution {
    public List<List<int>> list;
    public List<List<int>> Subsets(int[] nums) {
        list = new List<List<int>>();
        backtrack(nums, new List<int>(), 0);
        return list;
    }

    public void backtrack(int[] nums, List<int> curr, int index ){
        if (index == nums.Length){
            list.Add(new List<int>(curr));
            return;
        } 

        //dont add
        backtrack(nums, curr, index + 1);

        //add 
        curr.Add(nums[index]);
        backtrack(nums, curr, index + 1);
        curr.RemoveAt(curr.Count - 1); 
    }
}
